### PR TITLE
Adds functionality to ask the user to save before closing the editor

### PR DESCRIPTION
**Pull request by meetcshah19**
- Checks whether a level is open on pressing close if no then the editor quits.
- If yes then a prompt is displayed asking the user if he wants to save the level or not.

### CODE REVIEW

File: rootex/main/window.cpp [hunk 1]
```diff
@@ -2,23 +2,26 @@
 
 #include "input/input_manager.h"
 #include "renderer/rendering_device.h"
+#include "core/event_manager.h"
 
 #include "vendor/ImGUI/imgui.h"
 #include "vendor/ImGUI/imgui_impl_dx11.h"
 #include "vendor/ImGUI/imgui_impl_win32.h"
 
+
```
Comment (sin3point14) [inline rootex/main/window.cpp:11]: extra newline

File: rootex/main/window.cpp [hunk 2]
```diff
@@ -2,23 +2,26 @@
...
 void Window::show()
 {
 	ShowWindow(m_WindowHandle, SW_SHOW);
 }
 
-std::optional<int> Window::processMessages()
+    std::optional<int> Window::processMessages()
```
Comment (sin3point14) [inline rootex/main/window.cpp:17]: extra tab

File: rootex/main/window.cpp [hunk 3]
```diff
@@ -2,23 +2,26 @@
...
 {
 	MSG msg;
 	while (PeekMessage(&msg, nullptr, 0, 0, PM_REMOVE)) //non-blocking message retrieval
 	{
-		if (msg.message == WM_QUIT)
+		switch (msg.message)
```
Comment (sin3point14) [inline rootex/main/window.cpp:22]: handle WM_DESTROY as mentioned before

File: rootex/main/window.cpp [hunk 4]
```diff
@@ -78,6 +81,7 @@ void Window::clearUnboundTarget()
 	RenderingDevice::GetSingleton()->clearUnboundRenderTarget(0.0f, 0.0f, 0.0f);
 #endif // DEBUG
 #endif // ROOTEX_EDITOR
+
```
Comment (sin3point14) [inline rootex/main/window.cpp:84]: extra endline

File: rootex/main/window.h [hunk 5]
```diff
@@ -23,8 +23,9 @@ class Window
 	Window(const Window&) = delete;
 	Window& operator=(const Window&) = delete;
 	~Window() = default;
-
+	void static quit(HWND a);
```
Comment (sin3point14) [inline rootex/main/window.h:26]: why is this?

File: rootex/app/project_manager.h [hunk 6]
```diff
@@ -8,9 +8,9 @@ class ProjectManager
 	ProjectManager(ProjectManager&);
 	~ProjectManager() = default;
 
-	String m_CurrentLevelName;
```
Comment (murex971) [inline rootex/app/project_manager.h:11]: Remove extra line.

File: rootex/main/window.h [hunk 7]
```diff
@@ -10,21 +11,26 @@ class Window
...
 	int m_Height;
+	bool IsEditorWindow;
 
 	WNDCLASSEX m_WindowClass = { 0 };
 	LPCSTR m_ClassName;
 	HINSTANCE m_AppInstance;
 	HWND m_WindowHandle;
 	
 	static LRESULT CALLBACK WindowsProc(HWND windowHandler, UINT msg, WPARAM wParam, LPARAM lParam);
+
```
Comment (murex971) [inline rootex/main/window.h:22]: Remove this extra line.

File: rootex/main/window.h [hunk 8]
```diff
@@ -10,21 +11,26 @@ class Window
...
 	WNDCLASSEX m_WindowClass = { 0 };
 	LPCSTR m_ClassName;
 	HINSTANCE m_AppInstance;
 	HWND m_WindowHandle;
 	
 	static LRESULT CALLBACK WindowsProc(HWND windowHandler, UINT msg, WPARAM wParam, LPARAM lParam);
+	
 
 public:
+
```
Comment (murex971) [inline rootex/main/window.h:25]: Same as above.

File: rootex/app/project_manager.h [hunk 9]
```diff
@@ -8,9 +8,9 @@ class ProjectManager
...
 
 public:
+	String m_CurrentLevelName;
 	static ProjectManager* GetSingleton();
```
Comment (murex971) [inline rootex/app/project_manager.h:14]: Make this member variable private.

File: editor/editor.cpp [hunk 10]
```diff
@@ -209,6 +204,8 @@ void Editor::drawDefaultUI()
 					ImGui::SameLine();
 					if (ImGui::Button("Create"))
 					{
+						if (ProjectManager::GetSingleton()->isAnyLevelOpen())
+							saveAll(nullptr);
```
Comment (IronicallySerious) [inline editor/editor.cpp:208]: Use this format for all `if` statements. Don't remove curly braces, it helps reduce bugs in the future:
```cpp
if (condition)
{
    // do something
}
```

File: editor/editor.cpp [hunk 11]
```diff
@@ -220,6 +217,8 @@ void Editor::drawDefaultUI()
 					{
 						if (ImGui::MenuItem(levelName.string().c_str()))
 						{
+							if (ProjectManager::GetSingleton()->isAnyLevelOpen())
+								saveAll(nullptr);
```
Comment (IronicallySerious) [inline editor/editor.cpp:221]: Same here

File: editor/editor.h [hunk 12]
```diff
@@ -21,7 +21,7 @@ class HierarchyGraph;
 class Editor
 {
 	int m_DockSpaceID;
-
+	String menuAction = "";
```
Comment (IronicallySerious) [inline editor/editor.h:24]: m_MenuAction

File: rootex/main/window.cpp [hunk 13]
```diff
@@ -105,14 +108,16 @@ LRESULT CALLBACK Window::WindowsProc(HWND windowHandler, UINT msg, WPARAM wParam
...
+		switch (msg)
+		{
+		case WM_CLOSE:
+			EventManager::GetSingleton()->call("quitWindowRequest", "quitWindowRequest", 0);
+			return 0;
+		case WM_DESTROY:
+			PostQuitMessage(0);
+			return 0;
+		}
+
```
Comment (IronicallySerious) [inline rootex/main/window.cpp:120]: This entire thing is indented twice as much as it is required

File: rootex/main/window.cpp [hunk 14]
```diff
@@ -122,6 +127,8 @@ Window::Window(int xOffset, int yOffset, int width, int height, const String& ti
     : m_Width(width)
     , m_Height(height)
 {
+	BIND_EVENT_MEMBER_FUNCTION("quitWindowRequest", Window::quitWindow);
```
Comment (IronicallySerious) [inline rootex/main/window.cpp:130]: Event names should always be PascalCase

File: rootex/main/window.cpp [hunk 15]
```diff
@@ -183,7 +191,25 @@ Window::Window(int xOffset, int yOffset, int width, int height, const String& ti
 	applyDefaultViewport();
 }
 
+Variant Window::quitWindow(const Event* event) {
```
Comment (IronicallySerious) [inline rootex/main/window.cpp:194]: Wrong opening brace style

File: rootex/main/window.cpp [hunk 16]
```diff
@@ -183,7 +191,25 @@ Window::Window(int xOffset, int yOffset, int width, int height, const String& ti
...
+	if (this->IsEditorWindow)
```
Comment (IronicallySerious) [inline rootex/main/window.cpp:195]: You do not need `this` here.

File: rootex/main/window.cpp [hunk 17]
```diff
@@ -183,7 +191,25 @@ Window::Window(int xOffset, int yOffset, int width, int height, const String& ti
...
+	}
+	else
+	{
+		PostQuitMessage(0);
+	}
+	return true;
+}
+Variant Window::quitEditorWindow(const Event* event)
+{
+	DestroyWindow(this->getWindowHandle());
```
Comment (IronicallySerious) [inline rootex/main/window.cpp:207]: Same, do not use `this` when not using it is an option

File: rootex/main/window.cpp [hunk 18]
```diff
@@ -183,7 +191,25 @@ Window::Window(int xOffset, int yOffset, int width, int height, const String& ti
...
+	return true;
+}
+
 HWND Window::getWindowHandle()
 {
 	return m_WindowHandle;
 }
+
```
Comment (IronicallySerious) [inline rootex/main/window.cpp:215]: Delete this newline

File: rootex/main/window.h [hunk 19]
```diff
@@ -10,6 +11,7 @@ class Window
 protected:
 	int m_Width;
 	int m_Height;
+	bool IsEditorWindow;
```
Comment (IronicallySerious) [inline rootex/main/window.h:14]: m_IsEditorWindow

File: rootex/main/window.h [hunk 20]
```diff
@@ -20,11 +22,13 @@ class Window
 
 public:
 	Window(int xOffset, int yOffset, int width, int height, const String& title, bool isEditor, bool MSAA);
+	Variant quitWindow(const Event* event);
+	Variant quitEditorWindow(const Event* event);
```
Comment (IronicallySerious) [inline rootex/main/window.h:26]: Make these private. Probably shift up by 3 lines

File: rootex/main/window.h [hunk 21]
```diff
@@ -20,11 +22,13 @@ class Window
...
 	Window(const Window&) = delete;
 	Window& operator=(const Window&) = delete;
 	~Window() = default;
-
 	void show();
+
```
Comment (IronicallySerious) [inline rootex/main/window.h:31]: Don't need this change

File: rootex/main/window.h [hunk 22]
```diff
@@ -36,6 +40,5 @@ class Window
 	int getWidth() const;
 	int getHeight() const;
 	HWND getWindowHandle();
-
```
Comment (IronicallySerious) [inline rootex/main/window.h:39]: Don't need this change

File: editor/editor.h (hunk 12 above)
Comment (IronicallySerious) [reply editor/editor.h:24]: Also it is empty by default, no need to explicitly set it to empty string

**Pull request merged**